### PR TITLE
[jax2tf] Change the conversion of dot_general to use XLA op.

### PR DESCRIPTION
**Pull request by gnecula**
Instead of converting the dot_general to a sea of TF ops, when
we enable_xla we just use the XLA op. This has the advantage
that it also supports the preferred_element_type.

Also, it turns out that `tf.linalg.matmul` now supports tf.bfloat16 and tf.int32, so change the 
conversion for `not _enable_xla` to use `tf.linag.matmul` for those cases as well.

### CODE REVIEW

**Comment by gnecula** [comment]
Hold off reviewing, I found more problems.

File: jax/experimental/jax2tf/examples/serving/model_server_request.py
Comment (saeta) [inline jax/experimental/jax2tf/examples/serving/model_server_request.py:18]: For my own edification, could you explain why this is the right approach here?

**Comment by gnecula** [comment]
This is ready now. I will likely try to merge very soon, because it is blocking other changes. Please do send comments and I will incorporate them in a future CL.

File: jax/_src/lax/lax.py
Comment (saeta) [inline jax/_src/lax/lax.py:6655]: Doc string? (e.g. to explain the meaning of the tuple values)

File: jax/experimental/jax2tf/jax2tf.py
Comment (saeta) [inline jax/experimental/jax2tf/jax2tf.py:1239]: The third argument is surprising to me.

**Review by saeta — APPROVED** [review]
I feel somewhat less qualified to review this PR. I'm preemptively approving, but if you would like a more substantive review, I recommend either a video call, or someone else if you're more pressed for time.

File: jax/experimental/jax2tf/jax2tf.py
Comment (gnecula) [reply jax/experimental/jax2tf/jax2tf.py:1239]: The `*`? It says that the rest of the arguments must be passed as kwargs, not positional.

File: jax/experimental/jax2tf/examples/serving/model_server_request.py
Comment (gnecula) [reply jax/experimental/jax2tf/examples/serving/model_server_request.py:18]: We run `mypy` in OSS, and it complains if it cannot find one of the modules. `grpc` is not a standard requirement for JAX, and we don't want it to be. You need it only to run this serving example. For myself, when `mypy` complains, if I do not see the type error, I just `type: ignore`. My thinking is that we want the type annotations for documentation first, and the actual type checking is best effort.

**Comment by gnecula** [comment]
> I feel somewhat less qualified to review this PR. I'm preemptively approving, but if you would like a more substantive review, I recommend either a video call, or someone else if you're more pressed for time.

Indeed, there are many details here. I sent this to you mostly as education. I do think it is worth trying to digest it, but indeed it may be easier over a video call.

File: jax/_src/lax/lax.py
Comment (gnecula) [reply jax/_src/lax/lax.py:6655]: I added a docstring. I wish we did have more docstrings. For now, what we are trying to do is to have docstrings for all API functions, then ideally for all module exports, and probably not add them for module locals. I think that we should
add type annotations for all functions though.

**Pull request closed without merging**
